app/models/callback.rb

```diff
@@ -1,23 +1,30 @@
 class Callback < Base
   attribute :telephone, :string
-  # may need to DateTime.parse
   attribute :phone_call_scheduled_at, :string
 
   validates :telephone, length: { minimum: 5, too_short: "Telephone number is too short (minimum is 5 characters)" }, format: { with: /\A[0-9\s+]+\z/, message: "Enter a telephone number in the correct format" }
-  # possible caching issue
   validates :phone_call_scheduled_at, callbacks: { method: :get_callback_booking_quotas }
 
-  def self.options
-    quotas = ApiClient.get_callback_booking_quotas
-    grouped_quotas = quotas.group_by(&:day)
-    options_hash = Hash.new { |hash, key| hash[key] = [] }
-    grouped_quotas.each do |day, data|
-      data.each do |x|
-        gmt_start_slot = Time.zone.parse(x.start_at.to_s).strftime("%I:%M %P")
-        gmt_end_slot = Time.zone.parse(x.end_at.to_s).strftime("%I:%M %P")
-        options_hash[day] << [gmt_start_slot.to_s + " - " + gmt_end_slot.to_s, x.start_at]
+  class << self
+    def grouped_quotas
+      ApiClient.get_callback_booking_quotas.group_by(&:day)
+    end
+
+    def options
```

Comment:
This looks like view code and could be moved out to a helper method - you could have an option on the helper method to strip out today, than have your view opt in to it where appropriate - something like the following

```ruby
# my_data_model.rb
class MyDateModel
  def get_dates
    ApiClient.get_callback_booking_quotas 
   # massage this into nice structured data if not already
  end

  def get_dates_exc_today
    get_dates.reject { |d| d.day == Time.zone.today }
end

# booking_dates_helper.rb
def group_booking_dates(dates)
  # logic for grouping dates, and formatting them
  # See Ross' comments
end

# _my_form_partial.rb
f.grouped_collection_select :my_slot_attribute, 
    group_booking_dates(f.object.get_quotas_exc_today), 
    :first, 
    :last,
    :first,
    :last
```

Reply:
👍